bin/propolis-server/src/lib/migrate/destination.rs

```diff
@@ -24,105 +27,321 @@ use crate::migrate::probes;
+            EnsureState::Active(vm) => vm.state_publisher(),
+        };
+
+        publisher.update(ExternalStateUpdate::Migration(
+            MigrationStateUpdate { state, id, role: MigrateRole::Destination },
+        ));
+    }
+
+    /// Notifies the instance ensure helper that migration has failed and that
+    /// VM initialization should therefore also fail.
+    async fn fail(self, reason: MigrateError) -> MigrateError {
+        match self {
+            EnsureState::NotStarted(vm) => {
+                let reason = vm.fail(reason.into()).await;
+                reason
+                    .downcast::<MigrateError>()
+                    .expect("original error was a MigrateError")
+            }
+            EnsureState::Active(vm) => {
+                vm.fail().await;
+                reason
+            }
         }
-        return Err(err);
     }
 
-    Ok(())
+    /// If the state machine is in the `Active` state, locks the VM objects
+    /// shared and returns the corresponding guard. Returns `None` otherwise.
+    async fn vm_objects(&self) -> Option<VmObjectsShared> {
+        let objects = match self {
+            EnsureState::NotStarted(_) => return None,
+            EnsureState::Active(vm) => vm.vm_objects(),
+        };
+
+        Some(objects.lock_shared().await)
+    }
 }
 
-struct DestinationProtocol<T: AsyncRead + AsyncWrite + Unpin + Send> {
+/// The runner for version 0 of the LM protocol, using RON encoding.
+struct RonV0<'e, T: MigrateConn> {
+    /// The ID for this migration.
+    migration_id: Uuid,
+
     /// The logger for messages from this protocol.
     log: slog::Logger,
 
     /// The channel to use to send messages to the state worker coordinating
     /// this migration.
-    command_tx: tokio::sync::mpsc::Sender<MigrateTargetCommand>,
-
-    /// The channel that receives responses from the state worker coordinating
-    /// this migration.
-    response_rx: tokio::sync::mpsc::Receiver<MigrateTargetResponse>,
-
-    /// Transport to the source Instance.
     conn: WebSocketStream<T>,
 
     /// Local propolis-server address
     /// (to inform the source-side where to redirect its clients)
     local_addr: SocketAddr,
 
-    /// The VM objects into which to import the source VM's state. Only
-    /// initialized after the sync phase.
-    vm_objects: Option<Arc<crate::vm::objects::VmObjects>>,
+    /// The current state of the overall instance-ensure operation. Wrapped in
+    /// an `Option` to allow the various ensure state-types to be taken and
+    /// replaced.
+    ensure_state: Option<EnsureState<'e>>,
 }
 
-impl<T: AsyncRead + AsyncWrite + Unpin + Send> DestinationProtocol<T> {
+#[async_trait::async_trait]
+impl<'e, T: MigrateConn + Sync> DestinationProtocol<'e> for RonV0<'e, T> {
+    async fn run<'ensure>(
+        mut self,
+        ensure: VmEnsureNotStarted<'ensure>,
+    ) -> Result<VmEnsureActive<'ensure>, MigrateError>
+    where
```

Comment:
I'm not totally sure if this actually makes things all that much better or not (haven't actually tried it yet)  but I kind of wonder whether it would be a bit nicer if `ensure_state` were a local variable in this function, rather than a field on `RonV0`, and we passed it into the methods on `RonV0`. That way, I _think_ we could avoid `take`ing and `unwrap`ping it as often.

I might try sketching that out in a PR against this branch, if you're interested in seeing what I mean?

Reply:
Done in cbd51fa and 3c17194: I was worried this would end up being too verbose, but I think it's actually much simpler (and, taken to its conclusion, lets us get rid of the `EnsureState` enum entirely). Thanks for the suggestion!